cloudv2/counter.w

```diff
@@ -0,0 +1,237 @@
+
+  /// Returns the current value of the counter.
+  /// - `key` The key of the counter (defaults to "default")
+  inflight peek(key: str?): num;
+  
+  /// Sets the value of the counter.
+  /// - `value` The new value of the counter
+  /// - `key` The key of the counter (defaults to "default")
+  inflight set(value: num, key: str?): void;
+}
+
+pub class Counter impl ICounter {
+  inner: ICounter;
+  pub initial: num;
+  new(props: CounterProps) {
+    nodeof(this).title = "Counter";
+    nodeof(this).description = "A distributed atomic counter";
+    nodeof(this).icon = "calculator";
+    nodeof(this).color = "lime";
+
+    let target = util.env("WING_TARGET");
+    let id = nodeof(this).id;
+    this.initial = props.initial ?? 0;
+    if target == "sim" {
+      this.inner = new Counter_sim(props) as id;
+    } elif target == "tf-aws" {
+      this.inner = new Counter_tfaws(props) as id;
+    } else {
+      throw "Unsupported target: " + target;
+    }
+    nodeof(this.inner).hidden = true;
+
+    this.addUI();
+  }
+
+  addUI() {
+    new ui.Field("Actual value", inflight () => { return "{this.peek()}"; }) as "ValueField";
+    new ui.Button("Increment", inflight () => { this.inc(); }) as "IncrementButton";
+    new ui.Button("Decrement", inflight () => { this.dec(); }) as "DecrementButton";
+    new ui.Button("Reset", inflight () => { this.set(0); }) as "ResetButton";
+  }
+
+  pub inflight inc(amount: num?, key: str?): num {
+    return this.inner.inc(amount, key);
+  }
+
+  pub inflight dec(amount: num?, key: str?): num {
+    return this.inner.dec(amount, key);
+  }
+
+  pub inflight peek(key: str?): num {
+    return this.inner.peek(key);
+  }
+
+  pub inflight set(value: num, key: str?): void {
+    this.inner.set(value, key);
+  }
+}
+
+inflight class CounterBackend impl sim.IResource {
+  valuesFile: str;
+  initial: num;
+  statedir: str;
+  values: MutMap<num>;
+  new(ctx: sim.IResourceContext, initial: num) {
+    this.initial = initial;
+
+    this.statedir = ctx.statedir();
+    this.valuesFile = fs.join(this.statedir, "values.json");
+    if fs.exists(this.valuesFile) {
+      let data = fs.readJson(this.valuesFile);
+      // TODO: MutMap<num>.fromJson(...) - https://github.com/winglang/wing/issues/1796
+      this.values = unsafeCast(data["values"]);
+    } else {
+      this.values = {};
+    }
+  }
+
+  pub onStop() {
+    fs.writeJson(this.valuesFile, { "values": this.values.copy() });
+  }
+
+  pub inc(amount: num, key: str): num {
+    let prev = this.values.tryGet(key) ?? this.initial;
+    this.values[key] = prev + amount;
+    return prev;
+  }
+
+  pub dec(amount: num, key: str): num {
+    let prev = this.values.tryGet(key) ?? this.initial;
+    this.values[key] = prev - amount;
+    return prev;
+  }
+
+  pub peek(key: str): num {
+    return this.values.tryGet(key) ?? this.initial;
+  }
+
+  pub set(value: num, key: str) {
+    this.values[key] = value;
+  }
+}
+
+class Counter_sim impl ICounter {
```

Comment:
I would put each implementation in a separate file

Reply:
Done